nilearn/glm/first_level/first_level.py

```diff
@@ -959,107 +949,403 @@ def first_level_from_bids(dataset_path, task_label, space_label=None,
+    Filter to be used by :func:`get_bids_files`: \
+        :obj:`list` of :obj:`tuple` (str, str)
+        filters
+
+    """
+    filters = [('task', task_label)]
+
+    if space_label is not None:
+        filters.append(('space', space_label))
+
+    if extra_filter and supported_filters:
+        for filter_ in extra_filter:
+
+            if filter_[0] not in supported_filters:
+                warn(f"The filter {filter_} will be skipped. "
+                     f"'{filter_[0]}' is not among the supported filters. "
+                     f"Allowed filters include: {supported_filters}")
+                continue
+
+            filters.append(filter_)
+
+    return filters
+
+
+def _check_bids_image_list(imgs: list[str] | None,
+                           sub_label: str,
+                           filters: list[tuple[str, str]]) -> None:
+    """Check input BIDS images.
+
+    Check that:
+        - some images were found
+        - if more than one image was found, check that there is not more than
+          one image for a given session / run combination.
+
+    Parameters
+    ----------
+    imgs : list[str] or None
+        List of image fullpath filenames.
+
+    sub_label : :obj:`str`
+        Subject label as specified in the file names like _sub-<sub_label>_.
+
+    filters : :obj:`list` of :obj:`tuple` (str, str)
+        Filters of the form (field, label) used to select the files.
+        See :func:`get_bids_files`.
+
+    """
+    if not imgs:
+        raise ValueError('No bold files found '
+                         f'for subject {sub_label} '
+                         f'for filter: {filters}')
+
+    if len(imgs) <= 1:
+        return
+
+    msg_start = ("More than one nifti image found\n "
+                 f"for subject: '{sub_label}'\n"
+                 f"for filters: {filters}\n")
+    msg_end = ("Adapt that the arguments"
+               "'task_label', 'space_label' and 'img_filters'"
+               "to make sure only file per session, per run gets selected.")
+
+    run_check_list: list = []
+
+    for img_ in imgs:
+
+        parsed_filename = parse_bids_filename(img_)
+
+        if 'ses' in parsed_filename and 'run' in parsed_filename:
+            if ((parsed_filename['ses'], parsed_filename['run'])
+                 in set(run_check_list)):
+                raise ValueError(
+                    f"{msg_start}"
+                    f"for the same run {parsed_filename['run']} and "
+                    f"session {parsed_filename['ses']}. "
+                    f"{msg_end}")
+
+            run_check_list.append((parsed_filename['ses'],
+                                   parsed_filename['run']))
+
+        elif 'ses' in parsed_filename:
+            if parsed_filename['ses'] in set(run_check_list):
+                raise ValueError(
+                    f"{msg_start}"
+                    f"for the same ses {parsed_filename['ses']}, "
+                    f"while no additional run specification present. "
```

Comment:
No interpolation, so f-string is unnecessary. I believe flake8 flags this.

Reply:
done